src/java.base/share/classes/java/lang/ref/package-info.java

```diff
@@ -93,19 +93,57 @@
  * structure, this check will add little overhead to the hashtable
  * access methods.
  *
+ * <a id="MemoryConsistency"></a>
+ * <h3>Memory Consistency Properties</h3>
+ * Certain interactions between the garbage collector, references, and reference
+ * queues form
+ * <a href="{@docRoot}/java.base/java/util/concurrent/package-summary.html#MemoryVisibility"><i>happens-before</i></a>
+ * relationships:
+ *
+ * <ul>
+ *
+ * <li> Actions in a thread prior to calling
+ * {@link Reference#reachabilityFence Reference.reachabilityFence(x)}
+ * <i>happen-before</i> the garbage collector clears any reference to {@code x}.</li>
+ *
+ * <li>The clearing of a reference by the garbage collector <i>happens-before</i>
+ * the garbage collector enqueues the reference.</li>
+ *
+ * <li> The enqueueing of a reference (by the garbage collector, or
+ * by a successful call to {@link Reference#enqueue}) <i>happens-before</i>
+ * the reference is removed from the queue ("dequeued") by {@link ReferenceQueue#poll} or
+ * {@link ReferenceQueue#remove}.</li>
+ *
+ * <li>The dequeuing of a reference to a
+ * {@linkplain Cleaner#register(Object object, Runnable action) registered}
+ * object, by the Cleaner thread, <i>happens-before</i> the Cleaner thread runs
+ * the cleaning action for that object.</li>
+ *
+ * </ul>
+ *
+ * The above chain of <i>happens-before</i> edges ensures that actions in a
+ * thread prior to a {@link Reference#reachabilityFence Reference.reachabilityFence(x)}
+ * <i>happen-before</i> cleanup code for {@code x} runs on the cleaner thread.
+ *
+ * In particular, changes to the state of {@code x} made before
+ * {@code reachabilityFence(x)} will be visible to the cleanup code running on
+ * the cleaner thread without additional synchronization.
+ * See {@jls 17.4.5}.
+ *
  * <a id="reachability"></a>
  * <h3>Reachability</h3>
  *
- * Going from strongest to weakest, the different levels of
+ * A <em>reachable</em> object is any object that can be accessed in any potential
+ * continuing computation from any live thread (as stated in {@jls 12.6.1}).
```

Comment:
This seems like somewhat loose and sloppy wording to me. "Any potential continuing computation"? "Any live thread"? Could you share a pointer to JLS 12.6.1 being referenced here?

Reply:
https://docs.oracle.com/javase/specs/jls/se21/html/jls-12.html#jls-12.6.1

> A reachable object is any object that can be accessed in any potential continuing computation from any live thread. 

It may be "loose" because the devil is in the details when it comes to reachability, but I disagree that it is "sloppy". This expresses reachability in simple terms, as a "first-order" or "Newtonian" model. There are of course "Quantum" effects that need to be dealt with in practice. The JLS alludes to this with:
> Optimizing transformations of a program can be designed that reduce the number of objects that are reachable to be less than those which would naively be considered reachable.

Reply:
In fact, it appears as if the problem is with the use of "any", which is universal in strength, whereas the intention here is existential in strength (as suggested by. my wording). Indeed, you might achieve the same effect by replacing "any" with "some" so that:

"An object is reachable if it can be accessed in some continuing computation from some live thread."

You needn't even say live because dead threads can neither take steps nor continue participating in the computation nor can they "access" objects for whatever informal notion of access. The "some continuing computation" subsumes "potential" (as in a possible future) so potential can be dropped.

Reply:
I think you are overthinking this somewhat Ramki. I don't see a practical (non discrete-math) distinction between "some" and "any", so would not object to that single word change if it helps. But "potential" should remain as it covers branching in the program whereby if we proceed down one branch an object remains reachable, whereas if we precede down another then it may not.

Reply:
I don't think changing "any" to "some" is helpful. I think "any" is ambiguous regarding meaning universal or existential strength. The sense used here is, considering the possible future execution paths of a thread, if any of them accesses the object, that object is reachable. In other words, it means "any one" and not "all".

Reply:
OK, no worries; will let you decide what makes sense. Thanks!